File: 프로그래머스/2/68645. 삼각 달팽이/삼각 달팽이.java
class Solution {
    public int[] solution(int n) {
        
        int[][] snail = new int[n+1][n+1]; 
        
        int dirX[] = {1, 0, -1};
        int dirY[] = {0, 1, -1};
        int x = 0;
        int y = 0;
        int dir = 0;
        int num = 1;
        
        for(int i=n; i>0; i--){
            for(int j=0; j<i; j++){
                x += dirX[dir];
                y += dirY[dir];
                snail[x][y] = num;
                num++;
            }
            dir = (dir+1)%3;
        }
       
        int[] answer = new int[num-1];
        int index = 0;
        
        for(int i=1; i<n+1; i++){
            for(int j=0; j<i; j++){
                answer[index++] = snail[i][j];
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/2/68645. 삼각 달팽이/삼각 달팽이.js
function solution(n) {
    
    const snail = Array.from({length : n+1}, () => Array(n+1).fill(0));
    
    var dirX = [1, 0, -1];
    var dirY = [0, 1, -1];
    var dir = 0;
    var x = 0;
    var y = 0;
    var num = 1;
    
    for(let i=n; i>0; i--){
        for(let j=0; j<i; j++){
            x += dirX[dir];
            y += dirY[dir];
            snail[x][y] = num;
            num++;
        }
        dir = (dir+1)%3;
    }
    
    var answer = [];
    
    for(let i=1; i<n+1; i++){
        for(let j=0; j<i; j++){
            answer.push(snail[i][j]);
        }
    }
    
    return answer;
}